### ISSUE TITLE
Using symengine in C++

### ISSUE THREAD

**Issue by Polyacene** [symengine/symengine#1725]
Dear All,

I'd like to play with Symengine within a C++ project and couldn't find
overloaded operators for `+`, `*`, etc.  I therefore defined a few of them
in a test project. Is this the way to go or are these operators defined somehwere?
```
#include <symengine/basic.h>
#include <symengine/symbol.h>

#include <symengine/add.h>
#include <symengine/dict.h>
#include <symengine/integer.h>
#include <symengine/mul.h>
#include <symengine/pow.h>

#include <iostream>
#include <typeinfo> 

using SymEngine::RCP;
using SymEngine::symbol;
using SymEngine::add;
using SymEngine::mul;
using SymEngine::integer;

/// g++ symtest.C -o symtest -l symengine -l mpfr -l gmp -lflint && ./symtest  

std::ostream& operator<<(std::ostream& o, RCP<const SymEngine::Basic> &x)
{
    // o << x->__str__();
    o << *x;
    return o;
}

auto operator-(RCP<const SymEngine::Integer> y)
{
    return neg(y);
}

auto operator-(RCP<const SymEngine::Basic> y)
{
    return neg(y);
}

auto operator+(RCP<const SymEngine::Integer> x,  RCP<const SymEngine::Basic> y)
{
    return add(x, y);
}

auto operator+(RCP<const SymEngine::Basic> x,  RCP<const SymEngine::Integer> y)
{
    return add(x, y);
}

auto operator+(RCP<const SymEngine::Basic> x, RCP<const SymEngine::Basic> y)
{
    return add(x, y);
}

auto operator-(RCP<const SymEngine::Basic> x ,RCP<const SymEngine::Basic> y)
{
    return sub(x, y);
}

auto operator*(RCP<const SymEngine::Basic> x,RCP<const SymEngine::Basic> y)
{
    return mul(x, y);
}

int main()
{
    RCP<const SymEngine::Basic> t = symbol("t");
    RCP<const SymEngine::Basic> J = symbol("J");
    RCP<const SymEngine::Basic> U = symbol("U");
    
     std::cout << "t: " << t << " : " << typeid(t).name() << std::endl;
     
     auto y = t*t + U*J - t*U;
     auto z = - U + integer(2) + integer(3) * U;

     std::cout << "y: " << y << " : " << typeid(y).name()<< std::endl;
     std::cout << "z: " << z << " : " << typeid(z).name()<< std::endl;
    
}
```
In case this is the way to go, shouldn't that be part of symengine?

Best regards,
Peter

P.S.
I couldn't build the doc. make complains about a missing index.rst.

**Comment by certik** [comment]
Hi Peter, you can use the [Expression](https://github.com/symengine/symengine/blob/9d6a1ce14fa8716dded6aaa4e728248cfff5fdb5/symengine/expression.h#L28) class for this. The reason the core of SymEngine does not do that is that there might be a possible overhead and we did not want to complicate things with this extra layer. For end user code, if you prefer to use overloaded operators, you can just use the `Expression` class.

You can read more background information about this here:

https://github.com/symengine/symengine/blob/9d6a1ce14fa8716dded6aaa4e728248cfff5fdb5/doc/design.md#operator-overloading

**Comment by Polyacene** [comment]
Hi Čertík,
thanks for you response. The `Expression` class indeed looks nice.
I have to work now, but I'll look later into it. Since you mentioned possible
overhead, I'll do some benchmarks.

Thanks,
Peter

**Comment by Polyacene** [comment]
I've started to use symengine in some C++ test projects. I have to say that the lack of documentation makes ts a hurdle to get started. Nevertheless, I got little bit accustomed to it and start to really like it.

Concerning the overhead I tested the construction of Chebyshev Polynomials and I'm puzzled by the result. For small polynomials the fluctuations are larger than the timing differences.  However, for large polynomials `Expression` appears to be faster. In case you are interested, here's the code:
```
#include <chrono> 

#include <symengine/expression.h>


int main(const int argc, const char *argv[])
{
	int order = (argc > 1) ? atoi(argv[1]) : 5;
	
	auto x = SymEngine::symbol("x");
	
	///------------------------------
	/// Using Expression
	///------------------------------

	SymEngine::Expression T0("1"); 
	SymEngine::Expression T1(x); 
	auto xx = 2*T1;
	
	auto start = std::chrono::steady_clock::now();
	for( int n=2; n<=order; ++n)
	{
		T0 = xx * T1 - T0;
		std::swap(T0, T1);
		T1 = SymEngine::expand(T1);
	}
	auto end = std::chrono::steady_clock::now();
	std::chrono::duration<double> elapsed_seconds = end - start;
	std::cout << "Elapsed time Expression: " << elapsed_seconds.count() << std::endl;
		
	///------------------------------
	/// Using RCP
	///------------------------------
	
	SymEngine::RCP<const SymEngine::Basic> yy = SymEngine::mul(SymEngine::integer(2), x);
	SymEngine::RCP<const SymEngine::Basic> R0(SymEngine::integer(1));
	SymEngine::RCP<const SymEngine::Basic> R1(x);
	
	start = std::chrono::steady_clock::now();
	for( int n=2; n<=order; ++n)
	{
		R0 = SymEngine::sub(SymEngine::mul(yy, R1), R0);
		std::swap(R0, R1);
		R1 = SymEngine::expand(R1);
	}
	end = std::chrono::steady_clock::now();
	elapsed_seconds = end - start;
	std::cout << "Elapsed time RCP<>     : " << elapsed_seconds.count() << std::endl;

	//std::cout << "order: " << order << " :\t" << T1 << std::endl;
	//std::cout << "order: " << order << " :\t" << *R1 << std::endl;
}
```
and my timings:
```
SymEngine: ./symtime  10
Elapsed time Expression: 0.000155065
Elapsed time RCP<>     : 0.000113966

SymEngine: ./symtime  100
Elapsed time Expression: 0.00842689
Elapsed time RCP<>     : 0.00853987

SymEngine: ./symtime  1000
Elapsed time Expression: 0.330251
Elapsed time RCP<>     : 0.315595

SymEngine: ./symtime  5000
Elapsed time Expression: 11.2805
Elapsed time RCP<>     : 12.6148

SymEngine: ./symtime  5000
Elapsed time Expression: 11.2561
Elapsed time RCP<>     : 12.3774

SymEngine: ./symtime  5000
Elapsed time Expression: 11.421
Elapsed time RCP<>     : 12.5915
```

**Comment by certik** [comment]
Try to separate the benchmarks to separate files so that there is only one benchmark per file. The memory is typically trashed a bit, so the second benchmark always executes a bit slower in my experience.

**Comment by Polyacene** [comment]
Well spotted. I just exchanged the order of the tests, and the timings are now just the other way round.

**Comment by certik** [comment]
My theory is that the first benchmark does a lot of small allocations and deallocations and the malloc trashes memory somehow, so when you do it again with the second benchmark, it is not as fast, as malloc takes more time. I've seen this happen in my benchmarks. So one must be careful.

**Comment by Polyacene** [comment]
I didn't expect it as the memory footprint is tiny. In return, at least for this type of operations there's no performance penalty in using the nicer syntax via operator overloading.

**Comment by certik** [comment]
Yes, for this benchmark I would expect the overhead to be tiny, as most of the work is spent doing the actual operation `T0 = xx * T1 - T0` and `expand()`  on it.

To be sure there is always 0% overhead, we would need to write up usage on this class into tests, compile with all supported compilers and then examine the assembly code to ensure there is no overhead.

For your application I think you can safely use the `Expression` class.

Regarding documentation, we are working on it with @HaoZeke.

If you have any wishes for any API improvements, please let us know!

**Comment by isuruf** [comment]
Since there are no virtual methods defined on `Expression` class, I would expect 0% memory overhead.

**Comment by HaoZeke** [comment]
Hi, I @Polyacene; unfortunately, I haven't yet gotten around to updating this repository with the documentation details. We have the [C++ API docs here](https://symengine.org/api-docs/index.html); and [some general introductory stuff here](https://symengine.org/).

**Comment by Polyacene** [comment]
@isuruf Sometimes there are strange surprises. Therefore I wanted to perform at least a simple test.

@HaoZeke To be honest, a documentation as a list of the classes does not help much. My friend was `find` and `grep` on the source code and `typeid().name()`.  For me that's fine, as I'm used to C++. My co-worker, with whom I was looking into SymEngine, had a pretty hard time,  and I'm sure, he would have stopped without my help. More examples as in `firststeps.html` are in my opinion much better,
e.g. like the documentation of SymPy.

In the problem we are looking in, we generate a large set of equations from a C++ code, where we actually have many duplicates. `SymEngine::set_basic` is the perfect solution for the problem. However, you have to find it . If you are interested, I can try to setup examples based on our experience with symengine, which you may clean up and add to the documentation.
I know that writing documentation is lots of work, but it enhances  a library considerably.

@certik Concerning the API, I'll keep it in mind. I can't spent too much time on symengine currently, but I'll continue to use it and will provide my feedback. The use of `Expression` works pretty nicely. Already using `symbol` brings back the RCP objects. It's fine once one knows it, but for beginners it's irritating. Also, it's not clear, whether one should use `get_items()` or `get_value()`.

**Comment by HaoZeke** [comment]
@Polyacene it'd be absolutely fantastic to have some examples from you! Currently we are working on SymPy style introductions to the API, but real user use cases would be perfect. I'd definitely be happy to add them to the documentation with attribution.

**Comment by Lecrapouille** [comment]
It would be nice indeed to add documentation and C++ tutorials. I'm just starting with this lib because a friend of mine asked me else, no offend, I would not have given any effort trying to figure how to use it.

**Comment by HaoZeke** [comment]
@Lecrapouille it is important to understand where the confusion and complications are coming from. Are you confused about the usage of SymEngine in terms of it being a C++ library or confused in terms of using a symbolic math library? The former is covered by our API, and the latter, though important from a UX perspective, are perhaps covered better by a Computer Algebra System course.

**Comment by Lecrapouille** [comment]
@HaoZeke Yes your right: beginners are the best to explain the weak points of the project where they get glued. I follow @Polyacene it's hard to enter to a new project especially a project offering no general doc or tutorials.

I know CAS, I did it with my TI89 when I was younger and C++ is not an issue for me. I was testing if possible to use symbolic formulas in Forth (everse polish notation language).

My issue was my way of selecting a Github project: I'm always checking if the project contains a good readme file, then checks if the project contains examples/tutorials, then I look at issues and PRs (is the project is alive), finally if the code has comments. Finally, if it has, ideally, software architecture doc (very few of them has). If this is not the case, I keep searching for another Github project especially when seeing that user asked for documentation since 2016. I never read Doxygen API html pages because this just allows describing functions but not the goal of the project and I guess people need more software architecture (how functions interact together. A better alternative to Doxygen would be CWEB http://www.literateprogramming.com/cweb_download.html but this is another topic). What to look first on an unknown API ? The tutorial has to introduce main API functions, then if I'm looking for a specific feature I'll dive into the code.

Since my friend tried SymEngine I tried I spent hours figuring out to create and compile on my side a basic hello-world application in C++ and how to link it. I'm not using CMake and I noticed this project does not install pkg-config that would avoid me searching that I missed `-lgmp`. Just for a basic hello world, I had to dig into benchmarks folder. I did not try to look for if a derivative function existed. Just after this, I found https://www.ginac.de/ and I could write and run a C++ in 5 min thanks to the pkg-config. I noticed they have a good PDF describing the project. I spend a little time finding derivation but I get it. So I'll finally switch ginac (no offend). I really think that doc and examples are as important as the project itself. With documentation, I would not have find ginac (sadly they no longer use github which makes them difficult to find).

**Comment by Polyacene** [comment]
Sorry for my long silence there was just too much work.
I've not forgotten the examples, and I'll post one in another thread as Symengine sadly can't handle it.

@HaoZeke @Lecrapouille I agree that it was easier to start with ginac using the examples, but I found it more confusing than Symengine.
